Profiling tools (#187)
hpi-swa-lab/SqueakByExample-english: Environment/Environment.tex

SBEScreenshotRecorder writeTo: './figures/TimeProfile.png' building: [:helper |
	| sem window splitter list |
	sem := Semaphore new.
	[Processor activeProcess environmentAt: ActiveWorldVariable put: helper world. "avoid #useWorldAndHandDuring: to keep stack clean"
	[World imageForm] timeProfile.
	sem signal] fork.
	sem wait.
	window := helper foregroundWindow.
	helper scaleWindow: window extent: 580 px @ 425 px.
	splitter := window allMorphs detect: [:m |
	(m isKindOf: ProportionalSplitterMorph) and: [m splitsTopAndBottom]].
	splitter repositionBy: 0 px @ 110 px.
	helper selectAnyListIn: window at: '##Leaves##' andScrollIntoViewBy: 3.
	list := helper listMorphIncludingMatch: '##Leaves##' in: window.
	helper
		type: '>>frameAndFillRectangle:fillStyle:borderStyle:' into: list;
		keyStroke: list key: Character return.
]
\end{ExecuteSmalltalkScript}
\begin{ExecuteSmalltalkScript}
SBEScreenshotRecorder writeTo: './figures/SpaceTally.png' building: [:helper |
	(SpaceTally new depth: 2; spaceTally: (PackageInfo named: #Morphic) classes) exploreWithLabel: 'SpaceTally'.
	helper scaleWindow: helper foregroundWindow extent: 500 px @ 225 px.
]
\end{ExecuteSmalltalkScript}

> First, you can send the message \mthind{BlockClosure}{timeToRun} to a block to measure the time it takes to execute it:

Author: ```suggestion
First, you can send the message \mthind{BlockClosure}{timeToRun} to a block to measure the time in milliseconds it takes to execute it:

```

Alternatively, you can select a piece of code in a workspace, debugger, \autc, open the yellow button menu, and select \menu{tally it}.
If a single execution of the code is too fast to measure (confidently), you can also \emph{benchmark} it by sending \mthind{BlockClosure}{bench} or \mthind{BlockClosure}{benchFor:} to a block.
This will run the block repeatedly and compute a few statistics about the execution time.
\begin{code}{NB: CANNOT TEST}
[10 factorial] bench --> '14,500,000 per second. 68.8 nanoseconds per run. 0 % GC time.'
\end{code}
\index{benchmarks}

\begin{figure}[btp]
	\begin{minipage}[t]{0.48\textwidth}
		\ifluluelse
			{\includegraphics[width=\textwidth]{TimeProfile}}
			{\includegraphics[scale=0.7]{TimeProfile}}
		\label{fig:TimeProfile}
	\end{minipage}
	\hfill
	\begin{minipage}[t]{0.48\textwidth}
		\ifluluelse
			{\includegraphics[width=\textwidth]{SpaceTally}}
			{\includegraphics[scale=0.7]{exploreWithLabel: 'SpaceTally'}}
		\label{fig:SpaceTally}
	\end{minipage}
	\caption{A time profiler (left) and a space tally (right).}
\end{figure}

If you have discovered that your code is too slow, you might wonder which parts of your code are responsible for the slowness.

> For that, \sq provides the \emphind{message tally} toool.

Reviewer: toool -> tool

Author: I should fix my Grammarly extension 😅

You can start it to record the execution times for your code by selecting the code and choosing \menu{spy on it} from the context menu.
Alternatively, you can let it observe all processes in the image from the menu item \menu{world main docking bar~\go{} Apps~\go{} Message Tally}.

> Note that the message tally is a sampling profiler, so for detailed information, you will have to let your code run for some seconds (e.g. by running it 1000 times if it only takes little time).

Reviewer: vll: "little time" -> "short time"

Author: According to DeepL Write, Google, and Grammarly, "a little time" (but I missed the "a") is actually more popular than "short time". :)

```suggestion
Note that the message tally is a sampling profiler, so for detailed information, you will have to let your code run for some seconds (e.g. by running it 1000 times if it only takes a little time).
```

> Note that the message tally is a sampling profiler, so for detailed information, you will have to let your code run for some seconds (e.g. by running it 1000 times if it only takes a little time).

Author: ```suggestion
Note that the message tally is a sampling profiler, so for detailed information, you will have to let your code run for some seconds (\eg by running it 1000 times if it only takes a little time).
```

\paragraph{Lexicon}
Objects understand all messages for which their class or their superclasses have methods.
The system browser introduced earlier can only show you the methods of one class.
If you want to know all the methods an instance of a class understands you can open the \ind{lexicon}.
You can get the lexicon from the system browser by opening the yellow button menu for the class and selecting \menu{browse protocol (p)}.

\paragraph{Profiling tools}
Sometimes you might write code that is too slow.
For that, \sq provides a number of profiling tools to measure execution time and help you find bottlenecks.
\index{profiling tools}

\begin{ExecuteSmalltalkScript}
SBEScreenshotRecorder writeTo: './figures/TimeProfile.png' building: [:helper |
	| sem window splitter list |
	sem := Semaphore new.
	[Processor activeProcess environmentAt: ActiveWorldVariable put: helper world. "avoid #useWorldAndHandDuring: to keep stack clean"
	[World imageForm] timeProfile.
	sem signal] fork.
	sem wait.

> 	Transcript showln: 'then'.

Author: ```suggestion
```
